src/rpad/pybullet_envs/suction_gripper_v2.py

```diff
@@ -0,0 +1,248 @@
+
+            # Describe the contact point in the TIP FRAME.
+            base_link_pos, base_link_ori, _, _, _, _ = p.getLinkState(
+                bodyUniqueId=self.mount_id,
+                linkIndex=self.tip_link_index,
+                computeLinkVelocity=0,
+                physicsClientId=self.client_id,
+            )
+            T_world_tip = base_link_pos, base_link_ori
+            T_tip_world = p.invertTransform(*T_world_tip)
+            T_world_contact = contact_pos_on_A, base_link_ori
+            T_tip_contact = p.multiplyTransforms(*T_tip_world, *T_world_contact)
+
+            # Describe the contact point in the OBJ FRAME. Note that we use the
+            # orientation of the gripper
+            objcom_link_pos, objcom_link_ori, _, _, _, _ = p.getLinkState(
+                bodyUniqueId=obj_id,
+                linkIndex=contact_link,
+                computeLinkVelocity=0,
+                physicsClientId=self.client_id,
+            )
+            T_world_objcom = objcom_link_pos, objcom_link_ori
+            T_objcom_world = p.invertTransform(*T_world_objcom)
+            T_obj_contact = p.multiplyTransforms(*T_objcom_world, *T_world_contact)
+
+            # Short names so we can debug later (parent frame; child frame),
+            pfp, pfo = T_tip_contact
+            cfp, cfo = T_obj_contact
+
+            self.contact_const = p.createConstraint(
+                parentBodyUniqueId=self.mount_id,
+                parentLinkIndex=self.tip_link_index,
+                childBodyUniqueId=obj_id,
+                childLinkIndex=contact_link,
+                jointType=p.JOINT_FIXED,
+                jointAxis=(0, 0, 0),
+                # parentFramePosition=[0, 0, 0],
+                parentFramePosition=pfp,
+                parentFrameOrientation=pfo,
+                childFramePosition=cfp,
+                childFrameOrientation=cfo,
+                physicsClientId=self.client_id,
+            )
+            p.changeConstraint(
+                self.contact_const,
+                maxForce=self.constraint_force,
+                physicsClientId=self.client_id,
+            )
+            self.activated = True
+            self.contact_link_index = contact_link
+        else:
+            print("not in contact - no valid points to attach to.")
+
+    def release(self):
+        """This should deactivate the suction gripper."""
+        if self.contact_const is not None:
+            p.removeConstraint(self.contact_const, self.client_id)
+            self.contact_const = None
+        self.activated = False
+        self.contact_link_index = None
+
+    def get_move_fn(
+        self, goal_pos: npt.NDArray, goal_ori: npt.NDArray
+    ) -> Callable[[], Tuple[npt.NDArray, bool]]:
+        """Returns a function which, when called, gives you the next control to execute. This should happen
+        at every time step."""
+
+        def move_fn() -> Tuple[npt.NDArray, bool]:
```

Comment:
This and pull_fn are trivially implemented right now to just copy the original functionality from the demo env - technically set_pull_cmd isn't even applying motor control